src/nextjournal/clerk.clj

```diff
@@ -58,53 +58,75 @@
      {:result ret#
       :time-ms (/ (double (- (. System (nanoTime)) start#)) 1000000.0)}))
 
+(defn- wrapped-var [var]
+  {::wrapped-var var})
+
+(defn- lookup-cached-result [results-last-run introduced-var hash cas-hash visibility]
+  (try
+    (let [value (or (get results-last-run hash)
+                    (thaw-from-cas cas-hash))]
+      (when introduced-var
+        (intern *ns* (-> introduced-var symbol name symbol) value))
+      (wrapped-with-metadata (if introduced-var (wrapped-var introduced-var) value) visibility hash))
+    (catch Exception _e
+      ;; TODO better report this error, anything that can't be read shouldn't be cached in the first place
+      #_(prn :thaw-error e)
+      nil)))
+
+(defn- cachable-value? [value]
+  (not (or (fn? value)
+           (instance? clojure.lang.IDeref value)
+           (instance? clojure.lang.MultiFn value)
+           (instance? clojure.lang.Namespace value))))
+
+(defn- cache! [digest-file var-value]
+  (try
+    (spit digest-file (hash+store-in-cas! var-value))
+    (catch Exception _e
+      #_(prn :freeze-error e)
+      nil)))
+
+(defn- eval+cache! [form hash digest-file introduced-var no-cache? visibility]
+  (let [{:keys [result]} (time-ms (binding [config/*in-clerk* true] (eval form)))
+        var-value (cond-> result (var? result) deref)
+        no-cache? (or no-cache?
+                      (config/cache-disabled?)
+                      (view/exceeds-bounded-count-limit? var-value))]
+    (when (and (not no-cache?) (cachable-value? var-value))
+      (cache! digest-file var-value))
+    (let [blob-id (cond no-cache? (view/->hash-str var-value)
+                        (fn? var-value) nil
```

Comment:
why do functions get no hash, but other non-cachable things do get hashes?

Reply:
because Clojure functions cannot be (de-)serialized.